zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/CassandraSpanConsumer.java

```diff
@@ -0,0 +1,283 @@
+      if (span.isEmpty()) continue;
+      statements.add( // Allows lookup without the span name
+        storeTraceServiceSpanName(s.traceId(), service, "", ts_micro, ts_uuid, s.duration())
+      );
+    }
+    return new StoreSpansCall(statements, keys);
+  }
+
+  /** Store the span in the underlying storage for later retrieval. */
+  BoundStatement storeSpan(Span span, UUID ts_uuid) {
+    boolean traceIdHigh = !strictTraceId && span.traceId().length() == 32;
+
+    // start with the partition key
+    BoundStatement bound = bind(insertSpan)
+      .setUUID("ts_uuid", ts_uuid)
+      .setString("trace_id", traceIdHigh ? span.traceId().substring(16) : span.traceId())
+      .setString("id", span.id());
+
+    // now set the data fields
+    if (traceIdHigh) {
+      bound.setString("trace_id_high", span.traceId().substring(0, 16));
+    }
+    if (null != span.parentId()) {
+      bound.setString("parent_id", span.parentId());
+    }
+    if (null != span.kind()) {
+      bound.setString("kind", span.kind().name());
+    }
+    if (null != span.name()) {
+      bound.setString("span", span.name());
+    }
+    if (null != span.timestamp()) {
+      bound.setLong("ts", span.timestamp());
+    }
+    if (null != span.duration()) {
+      bound.setLong("duration", span.duration());
+    }
+    if (null != span.localEndpoint()) {
+      bound
+        .set("l_ep", new EndpointUDT(span.localEndpoint()), EndpointUDT.class)
+        .setString("l_service", span.localServiceName());
+    }
+    if (null != span.remoteEndpoint()) {
+      bound.set("r_ep", new EndpointUDT(span.remoteEndpoint()), EndpointUDT.class);
+    }
+    if (!span.annotations().isEmpty()) {
+      List<AnnotationUDT> annotations = span.annotations().stream()
+        .map(AnnotationUDT::new)
+        .collect(Collectors.toList());
+      bound.setList("annotations", annotations);
+    }
+    if (!span.tags().isEmpty()) {
+      bound.setMap("tags", span.tags());
+    }
+    Set<String> annotationQuery = CassandraUtil.annotationKeys(span);
+    if (!annotationQuery.isEmpty()) {
+      bound.setString("annotation_query", COMMA_JOINER.join(annotationQuery));
+    }
+    if (null != span.shared()) {
+      bound.setBool("shared", span.shared());
+    }
+    if (null != span.debug()) {
+      bound.setBool("debug", span.debug());
+    }
+    return bound;
+  }
+
+  BoundStatement storeTraceServiceSpanName(
+      String traceId,
+      String serviceName,
+      String spanName,
+      long ts_micro,
+      UUID ts_uuid,
+      Long duration) {
+
+    int bucket = durationIndexBucket(ts_micro);
+    BoundStatement bound = bind(insertTraceServiceSpanName)
+      .setString("trace_id", traceId)
+      .setString("service", serviceName)
+      .setString("span", spanName)
+      .setInt("bucket", bucket)
+      .setUUID("ts", ts_uuid);
+
+    if (null != duration) {
+      // stored as milliseconds, not microseconds
+      // round up to tens of milliseconds (or hundredths of seconds)
```

Comment:
line 229 needs to be removed. rounding to centiseconds no longer happens.
(after further benchmarking it was shown to be a bad idea)